Add automation for release preparation

htdvisser opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

This draft Pull Request adds a nice little script that I've been writing for automating the process of preparing releases. Using a script hopefully makes it faster and less error-prone.

Let me know what you think!

Commit:
api: Add ttnpb.FieldsWithoutPrefix

Commit:
api: Relax NS fieldmask rules

Commit:
api,dr: Return codec examples

Commit:
api,dr: Fix nested maps in codec examples

Commit:
api: Add codec ID and field mask paths for device repository codecs

Commit:
dr: Add field mask paths to Device Repository codec API

htdvisser committed:
dev: Add automation for release preparation

johanstokking commented on .github/ISSUE_TEMPLATE/release.md:
````diff
@@ -14,74 +14,8 @@ This is a checklist for releases. This is filled in by both the releaser and the
-    ```
-
-- [ ] Create a pull request targeting `v3.${minor}`.
+- [ ] Run `tools/bin/prepare_release minor` or `tools/bin/prepare_release patch`.
````
Add prerequisites; `gh`

johanstokking reviewed (commented):
Looks good to me. Maybe put the existing documentation in a separate Markdown document for reference and/or in case we need to do things manually still in some edge case.

KrishnaIyer reviewed (approved):
Great stuff

Commit:
ns: Support session and MAC state import

Commit:
dev: Update CHANGELOG

dependabot[bot] committed:
dev: Bump elliptic from 6.5.3 to 6.5.4

Bumps [elliptic](https://github.com/indutny/elliptic) from 6.5.3 to 6.5.4.
- [Release notes](https://github.com/indutny/elliptic/releases)
- [Commits](https://github.com/indutny/elliptic/compare/v6.5.3...v6.5.4)

Signed-off-by: dependabot[bot] <support@github.com>

Commit:
util: Add as service to js sdk

asmulko committed:
console: Add profile settings link to dropdown menu

Commit:
console: Extend console api definitions

Add application server methods to encode/decode downlink/uplink

Commit:
console: Add payload formatter testing form

Commit:
dev: Update locales

Commit:
ns: Ensure RX2 data rate index is supported by band

Commit:
ns,util: Only use session key ID for OTAA

Commit:
ns: Report invalid fields

Commit:
dev: Regenerate messages

Commit:
ns: Return error from AssertSetDevice always

Commit:
ns: Expand test coverage

Commit:
ns: Refactor DeviceRegistry Set for clarity

adriansmares committed:
cli: Fix oneof fields setup

Commit:
ns: Improve documentation

Commit:
ns: Remove redundant required field check

johanstokking committed:
Merge pull request #3982 from TheThingsNetwork/feature/3771-configure-peering

adriansmares committed:
Merge pull request #3990 from TheThingsNetwork/fix/3808-as-gs-vet

Fix go vet warnings

htdvisser committed:
api: Fix extraction of OrganizationOrUserIdentifiers log fields

johanstokking committed:
all: Fix Redis flake test

johanstokking committed:
as: Fix test flake

Commit:
cli: Avoid generating random session key ID

Commit:
ns: Avoid validating pending session when not actually set

Commit:
ns: Fix invalid fieldpath

Commit:
ns: Fix session/MAC state pair validations

Commit:
ns,util: Add testing utilities

Commit:
ns: Expand test coverage

Commit:
ns: Do not require DevEUI for multicast devices

htdvisser committed:
Merge pull request #3986 from TheThingsNetwork/feature/goreleaser-docker

htdvisser committed:
Merge pull request #3995 from TheThingsNetwork/fix/3981-extract-log-fields

Fix extraction of OrganizationOrUserIdentifiers log fields

htdvisser committed:
dev: Upgrade Go to 1.16

Commit:
dev: Regenerate JS SDK definitions

adriansmares committed:
Merge pull request #3989 from TheThingsNetwork/fix/cli-oneof-fields-os

Fix CLI oneof fields setup

adriansmares committed:
as: Expose stack domain as header

adriansmares committed:
as: Improve application packages scalability

htdvisser committed:
Merge pull request #3998 from TheThingsNetwork/feature/go-1.16

Upgrade Go to 1.16

htdvisser committed:
all: Use basic events PubSub instead of noop

KrishnaIyer committed:
api: Add Gateway Claiming messages

KrishnaIyer committed:
cli: Add gateway claiming commands

KrishnaIyer committed:
dev: Add jsSdk definitions

dependabot[bot] committed:
dev: Bump y18n from 3.2.1 to 3.2.2

Bumps [y18n](https://github.com/yargs/y18n) from 3.2.1 to 3.2.2.
- [Release notes](https://github.com/yargs/y18n/releases)
- [Changelog](https://github.com/yargs/y18n/blob/master/CHANGELOG.md)
- [Commits](https://github.com/yargs/y18n/commits)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
dev: Bump y18n from 4.0.0 to 4.0.1 in /sdk/js

Bumps [y18n](https://github.com/yargs/y18n) from 4.0.0 to 4.0.1.
- [Release notes](https://github.com/yargs/y18n/releases)
- [Changelog](https://github.com/yargs/y18n/blob/master/CHANGELOG.md)
- [Commits](https://github.com/yargs/y18n/commits)

Signed-off-by: dependabot[bot] <support@github.com>

adriansmares committed:
Merge pull request #4001 from TheThingsNetwork/feature/3896-app-packages-scalability

Improve application packages scalability

Commit:
ns: Avoid generating desired parameters for multicast

Commit:
ns: Allow partial MAC state specification

Commit:
ns: Fix switch case condition

Commit:
ns: Disallow custom channels for US-like bands

Disallow `factory_preset_frequncies` for bands with non-symmetric channels

Commit:
util: Add placeholder storage-db command for open-source

kschiffer committed:
Merge pull request #3908 from TheThingsNetwork/dependabot/npm_and_yarn/elliptic-6.5.4

build(deps): bump elliptic from 6.5.3 to 6.5.4

kschiffer committed:
Merge pull request #4004 from TheThingsNetwork/dependabot/npm_and_yarn/y18n-3.2.2

build(deps): bump y18n from 3.2.1 to 3.2.2

kschiffer committed:
Merge pull request #4005 from TheThingsNetwork/dependabot/npm_and_yarn/sdk/js/y18n-4.0.1

build(deps): bump y18n from 4.0.0 to 4.0.1 in /sdk/js

htdvisser committed:
ci: Fix credentials for pkg/blob tests

KrishnaIyer committed:
console: Add LNS API Key to gateway general settings page

adriansmares committed:
Merge pull request #3999 from TheThingsNetwork/feature/3795-network-domain

Expose stack domain as header

Commit:
api: Use struct for formatter example inputs

KrishnaIyer committed:
Merge pull request #4003 from TheThingsNetwork/feature/2381-gateway-claiming-oc

Add Gateway Claiming API and CLI commands

Commit:
dr: Unit tests for payload formatter examples

rvolosatovs committed:
Merge pull request #3680 from rvolosatovs/feature/ns-session-import

htdvisser committed:
Merge pull request #4002 from TheThingsNetwork/fix/default-events-pubsub

Use basic events PubSub instead of noop

htdvisser committed:
Merge pull request #4010 from TheThingsNetwork/fix/2903-blob-test-credentials

Fix credentials for pkg/blob tests

Commit:
console,account: Extend icons list

Commit:
all: Introduce ratelimit package and configuration

Commit:
all: Introduce rate limiting base configuration

Commit:
as: Rate limit MQTT connections and Webhooks

Commit:
gs: Rate limit gateway connections

Commit:
gcs: Rate limit Gateway Configuration Server and BasicStation CUPS

Commit:
all: Allow CORS for X-Rate-Limit-* headers

Commit:
all: Rate limit interop server

Commit:
all: Rate limit auth, metrics and health HTTP endpoints

Commit:
util: Update CHANGELOG

Commit:
Merge pull request #4008 from TheThingsNetwork/issue/3984-improve-storage-db-error-message

Add placeholder storage-db command for open-source

Commit:
all: Run go mod tidy

Commit:
Merge pull request #3920 from TheThingsNetwork/feature/rate-limiting

Rate Limiting

htdvisser committed:
all: Simplify usage of IDs when only strings are needed

KrishnaIyer committed:
console: Update formatting

KrishnaIyer committed:
console: Add validation to LBS LNS Secret

htdvisser committed:
Merge pull request #4013 from TheThingsNetwork/fix/id-stringer

Simplify usage of IDs when only strings are needed

adriansmares committed:
as: Do not error on unknown provider

Commit:
all: Rename rate configuration to MaxPerMin

johanstokking committed:
Merge pull request #4007 from rvolosatovs/fix/matching

Fix switch case condition

Commit:
console,account: Add field-group component

adriansmares committed:
Merge pull request #4015 from TheThingsNetwork/fix/as-provider-warning

Do not error on unknown PubSub provider

Commit:
all: Rename basicstation:cups rate limit class to http:gcs:cups

This is to enforce consistency with the rest of the configuration

Commit:
Merge pull request #4017 from TheThingsNetwork/quickfix/rate-limiting-config-name

Rate Limiting configuration quickfixes

asmulko committed:
Merge pull request #3968 from TheThingsNetwork/feature/3930-add-profile-settings-reference

Add profile settings link to dropdown menu

KrishnaIyer committed:
Merge pull request #4011 from TheThingsNetwork/feature/3929-lns-key-console

Add LNS API Key to gateway general settings page

KrishnaIyer committed:
data: Update external repositories

kschiffer committed:
account: Move code view to authenticated views

KrishnaIyer committed:
all: Bump to version 3.12.0

kschiffer committed:
dev: Update cypress tests

johanstokking committed:
Merge pull request #4022 from TheThingsNetwork/release/v3.12.0

Commit:
console: Remove dev_eui field for multicast devices

Commit:
console: Do not require ping slot periodicity for class c multicast

KrishnaIyer committed:
dev: Fix goreleaser npmfs config

KrishnaIyer committed:
dev: Update goreleaser dockers.binaries field

johanstokking committed:
Merge pull request #4025 from TheThingsNetwork/fix/go-releaser

Commit:
console: Fix lorawan version handling

KrishnaIyer committed:
dev: Update gorelease glob pattern

KrishnaIyer committed:
Merge pull request #4027 from TheThingsNetwork/fix/go-releaser-npfm

Update goreleaser glob pattern for nfpm

KrishnaIyer committed:
dev: Revert docker ids field to binaries

KrishnaIyer committed:
Merge pull request #4029 from TheThingsNetwork/revert/goreleaser-docker-ids

Revert docker ids field to binaries

Commit:
Merge pull request #4024 from TheThingsNetwork/fix/device-multicast

kschiffer committed:
Merge pull request #4023 from TheThingsNetwork/fix/authorization-count

Move authorization code view to authenticated views

htdvisser committed:
cli: Make list of allowed values in storage commands consistent

pgalic96 committed:
dev: Replace pgalic96 for asmulko in js codeowners

Commit:
api: Remove unused file

htdvisser committed:
api: Add explicit validators for request messages with embedded messages

pgalic96 committed:
Merge pull request #4031 from TheThingsNetwork/edit-codeowners

Replace pgalic96 by asmulko in js codeowners

Commit:
ns: Use uplink data rate index for ADR

htdvisser committed:
Merge pull request #4030 from TheThingsNetwork/fix/cli-storage-doc

Make list of allowed values in storage command flags consistent

htdvisser committed:
api: Improve validation of CreateAPIKeyRequests

Commit:
console,account: Fix byte input value handling in onChange

rvolosatovs committed:
Merge pull request #4034 from rvolosatovs/fix/adr

Use uplink data rate index for ADR

rvolosatovs committed:
Merge pull request #4033 from rvolosatovs/fix/cleanup

Remove unused testing file

pgalic96 committed:
console: Display None in the rights field if the API key has no rights

Commit:
console: Fix side navigation restricted entries

htdvisser committed:
is: Add options to reject common passwords

htdvisser committed:
Merge pull request #4035 from TheThingsNetwork/fix/create-api-key-validation

Improve validation of CreateAPIKeyRequests

htdvisser committed:
ci: Fix glob of public dir in tar.gz archives

htdvisser committed:
ci: Upload proto files to S3

htdvisser committed:
Merge pull request #4040 from TheThingsNetwork/fix/release-archives

Commit:
Merge pull request #4036 from TheThingsNetwork/fix/byte-onchange

Fix global `event` variable usage in byte input

johanstokking committed:
api: Add listed to PacketBrokerNetwork

johanstokking committed:
pba: Get and set whether the home network is listed

adriansmares committed:
as: Add required uplink message to tests

adriansmares committed:
console: Retrieve all webhook template fields

adriansmares committed:
console: Use fields default values

adriansmares committed:
api: Add optional field to webhook template fields

adriansmares committed:
console: Allow empty template fields

johanstokking committed:
Merge pull request #4043 from TheThingsNetwork/feature/pba-listed

Get and set listed in Packet Broker registration

adriansmares committed:
as: Return error on uplink simulation if skip payload crypto

Commit:
ns: Set FPort in tests

Commit:
ns: Fix `factory_preset_frequency` handling

- Assume `factory_preset_frequency` represents all enabled uplink
  channels on reset
- Merge with PHY channels instead of replacing

Commit:
util: Only print actual/expected in verbose tests

Commit:
api: Allow empty uplink frequency

Commit:
Merge pull request #4038 from TheThingsNetwork/fix/side-nav

Commit:
Merge pull request #4026 from TheThingsNetwork/fix/lw-version

Fix mac and phy version handling in the Console

pgalic96 committed:
Merge pull request #4037 from TheThingsNetwork/fix/empty-rights-api-key-display

Prevent API Key list views from crashing if the API key has no rights attached to it

Commit:
Merge pull request #3976 from TheThingsNetwork/feature/2232-payload-testing

htdvisser committed:
api: Skip validation of some empty fields

htdvisser committed:
is: Fix CUPS URI in Gateway test

htdvisser committed:
dev: Use dev protoc image

KrishnaIyer committed:
gs: Pass local variable ref to event call

htdvisser committed:
console: Close event stream when caller rights revoked

KrishnaIyer committed:
gs: Use different down instance in ScheduleDown

pgalic96 committed:
console: Add required fields to simulateUplink message

htdvisser committed:
Merge pull request #4050 from TheThingsNetwork/fix/event-stream-auth

Close event stream when caller rights revoked

KrishnaIyer committed:
Merge pull request #4049 from TheThingsNetwork/fix/gs-events-panic

Fix panic in Downlink event handling

htdvisser committed:
Merge pull request #4039 from TheThingsNetwork/fix/reject-common-passwords

Add options to reject common passwords

pgalic96 committed:
console: Make hardcoded fields consistent

adriansmares committed:
Merge pull request #4046 from rvolosatovs/feature/factory-preset-freq

Fix `factory_preset_frequency` handling

adriansmares committed:
Merge pull request #4044 from TheThingsNetwork/fix/webhook-templates-bugs

Webhook templates improvements

Commit:
dev: Add tippy dependency

Commit:
console,account: Add tooltip component

asmulko committed:
dev: Update jest dependency

asmulko committed:
Merge pull request #4054 from TheThingsNetwork/fix/update-jest-dependency

Update jest dependency

htdvisser committed:
data: Update external repositories

htdvisser committed:
Merge pull request #2669 from TheThingsIndustries/fix/request-validation

htdvisser committed:
all: Bump to version 3.12.1

Commit:
console: Add form field tooltips

Commit:
dev: Update locales

asmulko committed:
console: Add uint32 notation for device address

htdvisser committed:
Merge pull request #4060 from TheThingsNetwork/release/v3.12.1

adriansmares committed:
as: Break down the server interface

adriansmares committed:
as: Store uplink messages

adriansmares committed:
as: Add application uplinks storage

adriansmares committed:
as: Fix test usage of the application uplink storage

adriansmares committed:
ns: Allow LinkADRReq commands to lower the datarate

htdvisser committed:
all: Move ratelimit middleware after validation

adriansmares committed:
Merge pull request #4045 from TheThingsNetwork/feature/3913-more-skip-payload

Return error on uplink simulation if skip payload crypto

htdvisser committed:
Merge pull request #4064 from TheThingsNetwork/fix/grpc-middleware

adriansmares committed:
Merge pull request #4065 from TheThingsNetwork/fix/adr-up-down-379

Allow LinkADRReq commands to lower the datarate

adriansmares committed:
as: Add multi frame location requests API support

adriansmares committed:
as: Add multi frame location queries support

htdvisser committed:
ci: Revert uploading protos to S3

htdvisser committed:
Merge pull request #4070 from TheThingsNetwork/issue/cleanup-s3-protos

Revert uploading protos to S3

htdvisser committed:
api: Remove CombinedIdentifiers

htdvisser committed:
all: Use EntityIdentifiers in pkg/cluster

htdvisser committed:
all: Use EntityIdentifiers in pkg/events

htdvisser committed:
all: Use *ttnpb.EntityIdentifiers in pkg/auth/rights

htdvisser committed:
is: Use *ttnpb.EntityIdentifiers in MembershipStore

htdvisser committed:
is: Use *ttnpb.EntityIdentifiers in APIKeyStore

htdvisser committed:
cli: Stop using CombinedIdentifiers

htdvisser committed:
cli: Use GetEntityIdentifiers() instead of EntityIdentifiers()

htdvisser committed:
all: Stop using Identifiers()

htdvisser committed:
is: Use ttnpb.IDStringer in emails

htdvisser committed:
is: Remove use of EntityIdentifiers()

htdvisser committed:
all: Remove ttnpb.Identifiers from pkg/unique

htdvisser committed:
all: Remove ttnpb.Identifiers

dependabot[bot] committed:
dev: Update ssri dependency

htdvisser committed:
Merge pull request #4073 from TheThingsNetwork/issue/ttnpb-cleanup

asmulko committed:
Merge pull request #4074 from TheThingsNetwork/dependabot/npm_and_yarn/ssri-6.0.2

Update ssri package

Commit:
console: Add insufficient rights placeholder for root keysÑys

KrishnaIyer committed:
all: Add GCS to cluster config

asmulko committed:
Merge pull request #4058 from TheThingsNetwork/feature/4014-add-cstyle-uint32-option

Add uint32 notation for device address

KrishnaIyer committed:
dev: Update Changelog

adriansmares committed:
Merge pull request #4061 from TheThingsNetwork/feature/4018-recent-uplink-storage

Store recent data application uplinks in the Application Server

adriansmares committed:
as: Do not leak gateway identifiers to external services

adriansmares committed:
console: Expose multiframe configuration

adriansmares committed:
as: Add LoRaCloud GLS GNSS API

adriansmares committed:
as: Add LoRaCloud GLS GNSS queries support

Commit:
Merge pull request #4079 from TheThingsNetwork/fix/4071-key-rights

Show insufficient rights warning for root keys in the Console

Commit:
Merge pull request #4057 from TheThingsNetwork/feature/3809-context-popovers

KrishnaIyer committed:
Merge pull request #4080 from TheThingsNetwork/feature/add-gcs-to-cluster

Add GCS to cluster config

KrishnaIyer committed:
gcs: Handle requests with zero EUI

KrishnaIyer committed:
dev: Update changelog

KrishnaIyer committed:
dev: Update messages

adriansmares committed:
console: Add LoRaCloud GLS GNSS support

Commit:
api: Device Repository codec examples

Commit:
dr: Parse device repository codec examples

Commit:
dr: Unit tests for codec examples

Commit:
dr: Update device repository message processor

adriansmares committed:
console: Fix state updates

adriansmares committed:
as: Fix serialization time unit

adriansmares committed:
console: Add GLS WiFi support

adriansmares committed:
as: Add LoRaCloud GLS WiFi API support

adriansmares committed:
as: Add LoRaCloud GLS WiFi support

htdvisser committed:
all: Add validation to Echo forms

KrishnaIyer committed:
Merge pull request #4088 from TheThingsNetwork/fix/gcs-request-header

Handle requests with zero EUI

Commit:
account: Trim username on login page

Commit:
Merge pull request #4096 from TheThingsNetwork/fix/4048-login-username

Trim username when logging in

adriansmares committed:
as,console: Make naming convention consistent

adriansmares committed:
console: Reset form on token deletion

adriansmares committed:
Merge pull request #4068 from TheThingsNetwork/feature/4019-multiframe-request

LoRaCloud GLS multi frame request support

adriansmares committed:
Merge pull request #4089 from TheThingsNetwork/feature/4032-gnss

LoRaCloud GLS GNSS support

adriansmares committed:
Merge pull request #4093 from TheThingsNetwork/feature/4020-wifi-support

LoRaCloud GLS WiFi support

Commit:
api: Add validation to Device Repository examples

Commit:
dr: Refactor GetCodecRequest to an interface

htdvisser committed:
Merge pull request #4097 from TheThingsNetwork/fix/form-validation

all: Add validation to Echo forms

kschiffer committed:
console: Fix gateway downlink event display

kschiffer committed:
dev: Update event data for storybook

kschiffer committed:
dev: Update changelog

kschiffer committed:
Merge pull request #4101 from TheThingsNetwork/fix/4078-gs-downlink-events

Fix display of gateway downlink events in the Console

Commit:
dr: Fix incorrect handling of RX frequencies

Commit:
dr: Configure MaxEIRP for ABP devices only

Commit:
console: Do not set empty current parameters in device MAC state

Commit:
util: Tooling to update allowed field mask paths for JS SDK

Commit:
util: Update fixtures of device templates for E2E tests

Commit:
util: Update CHANGELOG

Commit:
Merge pull request #4086 from TheThingsNetwork/issue/console-enable-mac-state-import

Various fixes for end device creation from Console and Device Repository

Commit:
Merge pull request #3772 from TheThingsNetwork/feature/devicerepo-examples

Device Repository API for Codec Examples
